src/bentoml/_internal/env_manager/manager.py

```diff
@@ -0,0 +1,189 @@
+    ) as pyver_file:
+        py_version = pyver_file.read().split(".")[:2]
+        return ".".join(py_version)
+
+
+def run_script_subprocess(
+    script_file_path: t.Union[str, os.PathLike[str]],
+    capture_output: bool,
+    debug_mode: bool,
+):
+    shell_path = which("bash")
+    if shell_path is None:
+        raise BentoMLException("Unable to locate a valid shell")
+
+    safer_bash_args: list[str] = []
+
+    # but only work in bash
+    if debug_mode:
+        safer_bash_args = [
+            "-euxo",
+            "pipefail",
+        ]
+    result = subprocess.run(
+        [shell_path, *safer_bash_args, script_file_path],
+        capture_output=capture_output,
+    )
+    if result.returncode != 0:
+        logger.debug(decode(result.stdout))
+        logger.error(decode(result.stderr))
+        raise BentoMLException(
+            "Subprocess call returned non-zero value. Error: "
+            + decode(result.stderr)
+            + "\n Reffer logs for more details"
+        )
+
+
+@attrs.define
+class EnvManager:
+    env_name: str
+    env_type: str
+    is_ephimeral: bool
+    bento_path: t.Optional[str] = None
+    bentofile_path: t.Optional[str] = None
+    _env_fs: FS = attrs.field(init=False)
+
+    def __attrs_post_init__(self):
+        from bentoml._internal.configuration.containers import BentoMLContainer
+
+        env_home = fs.open_fs(BentoMLContainer.env_store_dir.get())
+
+        if not self.is_ephimeral:
+            assert (
+                self.env_name is not None
+            ), "persistent environments need a valid name."
+
+        assert not (
+            self.bento_path is None and self.bentofile_path is None
+        ), "Provide bentopath or path to bentofile."
+
+        if self.env_type == "conda":
+            if not env_home.exists("conda"):
+                env_home.makedir("conda")
+            self._env_fs = (
+                env_home.opendir("conda") if not self.is_ephimeral else TempFS()
+            )
+            self.create_conda_env()
+
+    @classmethod
+    def from_bento(
+        cls,
+        env_type: str,
+        is_ephimeral: bool,
+        env_name: t.Optional[str] = None,
+        bento_path: t.Optional[str] = None,
+    ) -> EnvManager:
+        if env_name is None:
+            env_name = "ephimeral_env"
+        return cls(
+            env_name=env_name,
+            env_type=env_type,
+            is_ephimeral=is_ephimeral,
+            bento_path=bento_path,
+        )
+
+    @classmethod
+    def from_bentofile(cls) -> EnvManager:
+        raise NotImplementedError
+
+    def create_conda_env(self) -> str:
+        """
+        Create a new conda env with self.env_name
+        """
+        conda_exe = os.environ.get("CONDA_EXE")
+        if conda_exe is None:
+            raise BentoMLException(
+                "conda executable not found! Make sure conda is installed and that `CONDA_EXE` is set."
+            )
```

Comment:
this is an excellent suggestion! I didn't realize `CONDA_EXE` was not universal, seems like only conda and miniconda implements it. but when I checked `CONDA_PREFIX` is also not universal and is not present if the env is deactivated in conda/miniconda. I think we might have to check both in this case.

maybe mambda/micromaba should be separate env as well. Because if all of them are installed, envManager will use the first one it gets, and no way for the user to configure that operation. That being said, I don't think that is a big problem either.